server/src/main/java/io/druid/server/http/security/RulesResourceFilter.java

```diff
@@ -0,0 +1,98 @@
+import com.google.common.collect.Iterables;
+import com.google.inject.Inject;
+import com.sun.jersey.spi.container.ContainerRequest;
+import io.druid.server.security.Access;
+import io.druid.server.security.AuthConfig;
+import io.druid.server.security.AuthorizationInfo;
+import io.druid.server.security.Resource;
+import io.druid.server.security.ResourceType;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.PathSegment;
+import javax.ws.rs.core.Response;
+
+
+/**
+ * Use this ResourceFilter when the datasource information is present after "rules" segment in the request Path
+ * Here are some example paths where this filter is used -
+ *  - druid/coordinator/v1/rules/
+ * */
+
+public class RulesResourceFilter extends AbstractResourceFilter
+{
+  @Inject
+  public RulesResourceFilter(AuthConfig authConfig)
+  {
+    super(authConfig);
+  }
+
+  @Override
+  public ContainerRequest filter(ContainerRequest request)
+  {
+    if (getAuthConfig().isEnabled()) {
+      // This is an experimental feature, see - https://github.com/druid-io/druid/pull/2424
+      final String dataSourceName = request.getPathSegments()
+                                           .get(
+                                               Iterables.indexOf(
+                                                   request.getPathSegments(),
+                                                   new Predicate<PathSegment>()
+                                                   {
+                                                     @Override
+                                                     public boolean apply(PathSegment input)
+                                                     {
+                                                       return input.getPath().equals("rules");
+                                                     }
+                                                   }
+                                               ) + 1
+                                           ).getPath();
+      Preconditions.checkNotNull(dataSourceName);
+      final AuthorizationInfo authorizationInfo = (AuthorizationInfo) getReq().getAttribute(AuthConfig.DRUID_AUTH_TOKEN);
+      Preconditions.checkNotNull(
+          authorizationInfo,
+          "Security is enabled but no authorization info found in the request"
+      );
+      final Access authResult = authorizationInfo.isAuthorized(
+          new Resource(dataSourceName, ResourceType.DATASOURCE),
+          getAction(request)
+      );
+      if (!authResult.isAllowed()) {
+        throw new WebApplicationException(
+            Response.status(Response.Status.FORBIDDEN)
+                    .entity(String.format("Access-Check-Result: %s", authResult.toString()))
+                    .build()
+        );
+      }
+    }
+
+    return request;
+  }
+
+  @Override
+  public boolean isApplicable(String requestPath)
+  {
+    return requestPath.startsWith("druid/coordinator/v1/rules/");
```

Comment:
Yes it will fail. However, as of now there is no end point which is exactly like `druid/coordinator/v1/rules/`. So, I am assuming whoever uses this ResourceFilter at a new endpoint will modify the logic so that it continues to work without failing with index out of bounds exception. Is this assumption reasonable ? otherwise what can be done here ?